lightning/src/ln/data_persister.rs

```diff
@@ -0,0 +1,98 @@
+//! Logic for persisting data from ChannelMonitors on-disk. Per-platform data
+//! persisters are separated into the lightning-persist-data crate. These
+//! objects mainly interface with the ChainMonitor when a channel monitor is
+//! added or updated, and when they are all synced on startup.
+use chain::channelmonitor::{ChannelMonitor, ChannelMonitorUpdate, ChannelMonitorUpdateErr};
+use chain::keysinterface::ChannelKeys;
+use chain::transaction::OutPoint;
+use std::collections::HashMap;
+
+/// ChannelDataPersister is responsible for persisting channel data: this could
+/// mean writing once to disk, and/or uploading to several backup services.
+///
+/// Note that for every new monitor, you **must** persist the new ChannelMonitor
+/// to disk/backups. And, on every update, you **must** persist either the
+/// ChannelMonitorUpdate or the updated monitor itself. Otherwise, there is risk
+/// of situations such as revoking a transaction, then crashing before this
+/// revocation can be persisted, then unintentionally broadcasting a revoked
+/// transaction and losing money. This is a risk because previous channel states
+/// are toxic, so it's important that whatever channel state is persisted is
+/// kept up-to-date.
+///
+/// Upon calls to update_channel_data, implementers of this trait should either
+/// persist the ChannelMonitorUpdate or the ChannelMonitor. If an implementer
+/// chooses to persist the updates only, they need to make sure that all the
+/// updates are applied to the ChannelMonitors *before* the set of channel
+/// monitors is given to the ChainMonitor at startup time (e.g., all monitors
+/// returned by `load_channel_data` must be up-to-date). If full ChannelMonitors
+/// are persisted, then there is no need to persist individual updates.
+///
+/// Note that there could be a performance tradeoff between persisting complete
+/// channel monitors on every update vs. persisting only updates and applying
+/// them in batches.
+///
+/// Given multiple backups, situations may arise where one or more backup sources
+/// have fallen behind or disagree on the current state. Because these backup
+/// sources don't have any transaction signing/broadcasting duties and are only
+/// supposed to be persisting bytes of data, backup sources may be considered
+/// "in consensus" if a majority of them agree on the current state and are on
+/// the highest known commitment number. See individual methods for more info.
+pub trait ChannelDataPersister: Send + Sync {
```

Comment:
Hm, should we factor in that we intend to persist other data through this module in the future as well (e.g. `ChannelManager`s, `Channel`s, etc)? Maybe it should just be called the `Persister`?

Reply:
I would say that `ChannelMonitor` and `Channel` should be persisted atomically together, otherwise inconsistency could result.  So that could be part of this trait.

Perhaps the non-channel-specific things in `ChannelManager` should be persisted using a different trait.  That could be named e.g. `NodePersister` or such, since it includes non-channel relevant things like the announcement serial number.

Reply:
That's a good point. The thing is that iiuc, stuff that needs to be persisted is pretty much divided into `ChannelMonitor`s (which need to be time sensitively persisted whenever a channel is updated) and everything else, which should be persisted sometimes and would suck to lose but isn't at the same level of sensitivity. 

How often the user wants to persist the less-sensitive stuff is subjective. We could make that decision for them (like the `ChannelDataPersister` does) but different users may have different fault-tolerance preferences. Not sure what the solution is here.

Reply:
Ok discussed offline, update:

There's only 2 structs that need to be persisted: (1) the `ChannelMonitor`s, and (2) the `ChannelManager`. The `ChannelMonitor`s are sensitive data, and them being out of date could result in funds lost. However, the `ChannelManager` is not sensitive, and if e.g. a node starts up with an out of date `ChannelManager`, this will result in channels being force closed, but there's no chance for loss of funds.

So basically this PR is gonna focus on persisting `ChannelMonitor`s, and we'll figure out how `ChannelManager` persistence is gonna work later. And the remaining open question is what to name the `ChannelDataPersister`.

Reply:
I see several fields in `Channel` that might be critical for correct functioning so need persistence, including:

- `channel_state`
- `channel_outbound`
- `value_to_self_msat`
- `short_channel_id` (also needed to restore `ChannelHolder` correctly)

and maybe others.  They probably won't cause loss of funds if out of date, but the system just won't work when restarted and restored from disk.

Also, `ChannelHolder` might have information that is required in order to be able to claim HTLCs, so if those are lost we can lose the HTLC value.

If we lose `ChannelManager.last_node_announcement_serial`, what is the impact?

Lastly, the `KeysManager` inside the `ChannelManager` has some state, including `channel_child_index` and `rand_bytes_child_index`.  I'm not sure what is the impact if we end up with the same keyset for two channels.

Reply:
> and maybe others. They probably won't cause loss of funds if out of date, but the system just won't work when restarted and restored from disk.

You need to store the whole channel policy as announced to your counterparty, including its own policy, otherwise you may have channel policy discrepancies and your counterparty closing the channel unexpectedly (e.g a `max_accepted_htlcs` violation). 

> Also, ChannelHolder might have information that is required in order to be able to claim HTLCs, so if those are lost we can lose the HTLC value.

Good point, you need to store HTLC-relay data too, otherwise you may lost an in-flight `update_fail_htlc` form an downstream channel to an upstream one.

I don't think we need to store the Events as they should be re-generated if we re-apply the state correctly to every component. 

> If we lose ChannelManager.last_node_announcement_serial, what is the impact?

BOLT7:

if node_id is NOT previously known from a channel_announcement message, OR if timestamp is NOT greater than the last-received node_announcement from this node_id: SHOULD ignore the message.

> Lastly, the KeysManager inside the ChannelManager has some state, including channel_child_index and rand_bytes_child_index. I'm not sure what is the impact if we end up with the same keyset for two channels.

Fund safety, so we should be careful storing it.